src/multicore.rs

```diff
@@ -1,81 +1,129 @@
+            // limits, we run the risk of memory exhaustion due to limited
+            // stack space consumed by all of the pending closures to be
+            // executed.
+            let previous_count = WORKER_SPAWN_COUNTER.fetch_add(1, Ordering::SeqCst);
+
+            // If the number of spawns requested has exceeded the number
+            // of cores available for processing by some factor (the
+            // default being 4), instead of requesting that we spawn a new
+            // thread, we instead execute the closure in the context of an
+            // install call to help clear the growing work queue and
+            // minimize the chances of memory exhaustion.
+            if previous_count > *WORKER_SPAWN_MAX_COUNT {
+                THREAD_POOL.install(move || {
+                    trace!("[{}] switching to install to help clear backlog[current threads {}, threads requested {}]",
+                           thread_index,
+                           THREAD_POOL.current_num_threads(),
+                           WORKER_SPAWN_COUNTER.load(Ordering::SeqCst));
+                    let res = f();
+                    sender.send(res).unwrap();
+                    WORKER_SPAWN_COUNTER.fetch_sub(1, Ordering::SeqCst);
+                });
+            } else {
+                THREAD_POOL.spawn(move || {
+                    let res = f();
+                    sender.send(res).unwrap();
+                    WORKER_SPAWN_COUNTER.fetch_sub(1, Ordering::SeqCst);
+                });
             }
+
+            Waiter { receiver }
         }
 
         pub fn scope<'a, F, R>(&self, elements: usize, f: F) -> R
         where
-            F: FnOnce(&Scope<'a>, usize) -> R,
+            F: FnOnce(&rayon::Scope<'a>, usize) -> R + Send,
+            R: Send,
         {
-            let chunk_size = if elements < self.cpus {
+            let chunk_size = if elements < *NUM_CPUS {
                 1
             } else {
-                elements / self.cpus
+                elements / *NUM_CPUS
             };
 
-            // TODO: Handle case where threads fail
-            crossbeam::scope(|scope| f(scope, chunk_size))
-                .expect("Threads aren't allowed to fail yet")
+            THREAD_POOL.scope(|scope| f(scope, chunk_size))
         }
     }
 
-    pub struct WorkerFuture<T, E> {
-        future: CpuFuture<T, E>,
+    pub struct Waiter<T> {
+        receiver: Receiver<T>,
     }
 
-    impl<T: Send + 'static, E: Send + 'static> Future for WorkerFuture<T, E> {
-        type Item = T;
-        type Error = E;
+    impl<T> Waiter<T> {
+        /// Wait for the result.
+        pub fn wait(&self) -> T {
+            if THREAD_POOL.current_thread_index().is_some() {
+                // Calling `wait()` from within the worker thread pool can lead to dead logs
+                error!("The wait call should never be done inside the worker thread pool");
+                debug_assert!(false);
+            }
```

Comment:
```suggestion
            assert!(
                THREAD_POOL.current_thread_index().is_none(),
                "wait() cannot be called from within the worker thread pool since that would lead to deadlocks",
            );
```

Reply:
The `error!` call here is logging, so I don't want to remove it if possible. I suspect the reason this uses a debug assertion is that

Reply:
The PR code doesn't gain any performance from using a debug assertion. The call to `THREAD_POOL.current_thread_index()` is done unconditionally. I suspect this is a policy difference, and our policy is for assertions to not depend on debug mode.

```suggestion
            if THREAD_POOL.current_thread_index().is_some() {
                let msg = "wait() cannot be called from within the worker thread pool since that would lead to deadlocks";
                error!(msg);
                panic!(msg);
            }
```

Reply:
The performance is not from avoiding the call to `THREAD_POOL.current_thread_index()`, but from continuing on and calling `self.receiver.recv()`. But thinking about this more, doing so should be a programming error, so I agree with this being an unconditional `panic!`.